### PR TITLE
Feature/multiassets chainlist

### PR DESCRIPTION
**Pull request by bnsports**
raw version of main screen. no on-fly updates there

### CODE REVIEW

File: fearless/Common/DataProvider/Subscription/PriceLocalStorageSubscriber.swift [hunk 1]
```diff
@@ -24,7 +24,7 @@ extension PriceLocalStorageSubscriber {
         }
 
         let options = DataProviderObserverOptions(
-            alwaysNotifyOnRefresh: false,
+            alwaysNotifyOnRefresh: true,
```
Comment (posplaw) [inline fearless/Common/DataProvider/Subscription/PriceLocalStorageSubscriber.swift:27]: Please leave comment to return when live updates come back

File: fearless/Common/Extension/Foundation/Decimal+String.swift [hunk 2]
```diff
@@ -7,4 +7,19 @@ extension Decimal {
 
         return NSDecimalString(&value, separator)
     }
+
+    func toString(digits _: Int) -> String? {
```
Comment (posplaw) [inline fearless/Common/Extension/Foundation/Decimal+String.swift:11]: why not digits: Int = 2 ?

File: fearless/Common/Helpers/AssetBalanceDisplayInfo.swift [hunk 3]
```diff
@@ -22,6 +22,17 @@ extension AssetBalanceDisplayInfo {
         )
     }
 
+    static func percent() -> AssetBalanceDisplayInfo {
+        AssetBalanceDisplayInfo(
+            displayPrecision: 2,
```
Comment (posplaw) [inline fearless/Common/Helpers/AssetBalanceDisplayInfo.swift:27]: Constants pls

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListInteractor.swift [hunk 4]
```diff
@@ -0,0 +1,121 @@
+import UIKit
+import RobinHood
+
+final class ChainAccountBalanceListInteractor {
+    weak var presenter: ChainAccountBalanceListInteractorOutputProtocol!
```
Comment (posplaw) [inline fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListInteractor.swift:5]: no force unwraps please

Comment (posplaw) [reply fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListInteractor.swift:5]: make optional, assert each time within guard let else or before calling presenter

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListInteractor.swift [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+extension ChainAccountBalanceListInteractor: ChainAccountBalanceListInteractorInputProtocol {
+    func setup() {
+        fetchChainsAndSubscribeBalance()
+
+        presenter.didReceiveSelectedAccount(selectedMetaAccount)
+    }
+
+    func refresh() {
+        if let priceProviders = priceProviders {
```
Comment (posplaw) [inline fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListInteractor.swift:105]: `priceProviders?.forEach { $0.refresh }` is MUCH lighter

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        guard
+            let asset = chain.utilityAssets().first,
+            let accountInfoResult = accountInfoResults[chain.chainId],
+            case let .success(accountInfo) = accountInfoResult else {
+            return nil
+        }
+
+        let assetInfo = asset.displayInfo
+
+        let maybeBalance: Decimal?
```
Comment (posplaw) [inline fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift:40]: what is weird name, maybe balance??

Comment (posplaw) [reply fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift:40]: just leave balance

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        if let accountInfo = accountInfo {
+            maybeBalance = Decimal.fromSubstrateAmount(
+                accountInfo.data.available,
+                precision: assetInfo.assetPrecision
+            )
+        } else {
+            maybeBalance = 0.0
+        }
+
+        guard let balance = maybeBalance else {
```
Comment (posplaw) [inline fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift:51]: simply `return balance?.stringwithPointSeparator`

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        let usdTokenFormatter = assetBalanceFormatterFactory.createTokenFormatter(for: usdDisplayInfo)
+        let usdTokenFormatterValue = usdTokenFormatter.value(for: selectedLocale)
+
+        guard let priceId = asset.priceId, let priceData = try? priceResults[priceId]?.get(),
+              let priceDecimal = Decimal(string: priceData.price) else {
+            return nil
+        }
+
+        let changeString: String = priceData.usdDayChange.map {
+            let percentValue = $0 / 100
```
Comment (posplaw) [inline fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift:69]: what is $0 ? optional double?

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+            return percentValue.percentString() ?? ""
+        } ?? ""
+
+        let priceString: String = usdTokenFormatterValue.stringFromDecimal(priceDecimal) ?? ""
```
Comment (posplaw) [inline fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift:73]: please remove `: String`

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+        let priceWithChangeString = [priceString, changeString].joined(separator: " ")
+
+        let priceWithChangeAttributed = NSMutableAttributedString(string: priceWithChangeString)
+
+        let color = (priceData.usdDayChange ?? 0) > 0 ? R.color.colorGreen() : R.color.colorRed()
```
Comment (posplaw) [inline fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift:79]: red for 0? maybe grey? please ask designer if not on design

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+
+    private func provideViewModel() {
+        let usdDisplayInfo = AssetBalanceDisplayInfo.usd()
+        let usdTokenFormatter = assetBalanceFormatterFactory.createTokenFormatter(for: usdDisplayInfo)
+        let usdTokenFormatterValue = usdTokenFormatter.value(for: selectedLocale)
+
+        var totalWalletBalance: Decimal = chainModels.compactMap { chainModel in
+
+            chainModel.assets.compactMap {
+                let balance = getBalance(for: chainModel) ?? ""
```
Comment (posplaw) [inline fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift:124]: why `?? ""` ? maybe return nil immediately?

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+
+        return usdTokenFormatterValue.stringFromDecimal(totalBalanceDecimal)
+    }
+
+    private func provideViewModel() {
+        let usdDisplayInfo = AssetBalanceDisplayInfo.usd()
+        let usdTokenFormatter = assetBalanceFormatterFactory.createTokenFormatter(for: usdDisplayInfo)
+        let usdTokenFormatterValue = usdTokenFormatter.value(for: selectedLocale)
+
+        var totalWalletBalance: Decimal = chainModels.compactMap { chainModel in
```
Comment (posplaw) [inline fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift]: maybe instead `compactMap { compactMap { ... } }` do `.reduce([], { $0 + $1.chainModel.assets }.compactMap { ... }` with merged array?

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+                      let priceDecimal = Decimal(string: priceData.price),
+                      let balanceDecimal = Decimal(string: balance) else {
+                    return nil
+                }
+
+                return priceDecimal * balanceDecimal
+            }.reduce(0, +)
+        }.reduce(0, +)
+
+        viewModels = chainModels.map { chainModel in
```
Comment (posplaw) [inline fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift:137]: maybe instead `compactMap { compactMap { ... } }` do `.reduce([], { $0 + $1.chainModel.assets }.compactMap { ... }` with merged array? chain-way is preferred

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListViewController.swift [hunk 14]
```diff
@@ -0,0 +1,121 @@
+import UIKit
+import SoraFoundation
+
+final class ChainAccountBalanceListViewController: UIViewController, ViewHolder {
+    typealias RootViewType = ChainAccountBalanceListViewLayout
```
Comment (posplaw) [inline fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListViewController.swift:5]: what is `Type` here for? no needed definitely

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListViewController.swift [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+
+    func configure() {
+        rootView.tableView.registerClassForCell(ChainAccountBalanceTableCell.self)
+
+        rootView.tableView.tableFooterView = UIView()
+
+        rootView.tableView.dataSource = self
+        rootView.tableView.delegate = self
+
+        if let refreshControl = rootView.tableView.refreshControl {
```
Comment (posplaw) [inline fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListViewController.swift:41]: why not `rootView.tableView.refreshControl?.addTarget` ?

File: fearless/Modules/MainTabBar/MainTabBarPresenter.swift [hunk 16]
```diff
@@ -26,9 +26,7 @@ extension MainTabBarPresenter: MainTabBarInteractorOutputProtocol {
         crowdloanListView = wireframe.showNewCrowdloan(on: view) as? UINavigationController
     }
 
-    func didUpdateWalletInfo() {
-        wireframe.reloadWalletContent()
-    }
+    func didUpdateWalletInfo() {}
```
Comment (posplaw) [inline fearless/Modules/MainTabBar/MainTabBarPresenter.swift:29]: If removed not for eternity, please leave comment

**Review by posplaw — CHANGES_REQUESTED** [review]

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListViewController.swift (hunk 14 above)
Comment (bnsports) [reply fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListViewController.swift:5]: @posplaw please explain. this class conforms to protocol ViewHolder that have associatedtype

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift (hunk 13 above)
Comment (bnsports) [reply fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift:137]: @posplaw the chainModel argument from first compactMap is used inside block. just array of assets not enough to build necessary objects

File: fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift (hunk 12 above)
Comment (bnsports) [reply fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift]: @posplaw the chainModel argument from first compactMap is used inside block. just array of assets not enough to build necessary objects

**Pull request closed without merging**